src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -708,14 +653,69 @@ internal override SyntaxNode ForbiddenZone
                             return forStatement.Deconstruction;
 
                         case SyntaxKind.ForEachComponentStatement:
-                            var forEachStatement = (ForEachComponentStatementSyntax)_deconstruction;
-                            return forEachStatement.Expression;
+                            // There is no forbidden zone for a foreach statement, because the
+                            // variables are not in scope in the expression.
+                            return null;
 
                         default:
                             throw ExceptionUtilities.UnexpectedValue(_deconstruction.Kind());
                     }
                 }
             }
         }
+
+        private class LocalSymbolWithEnclosingContext : SourceLocalSymbol
+        {
+            private readonly SyntaxNode _forbiddenZone;
+            private readonly Binder _nodeBinder;
+            private readonly SyntaxNode _nodeToBind;
+
+            public LocalSymbolWithEnclosingContext(
+                Symbol containingSymbol,
+                Binder scopeBinder,
+                Binder nodeBinder,
+                TypeSyntax typeSyntax,
+                SyntaxToken identifierToken,
+                LocalDeclarationKind declarationKind,
+                SyntaxNode nodeToBind,
+                SyntaxNode forbiddenZone)
+                : base(containingSymbol, scopeBinder, false, typeSyntax, identifierToken, declarationKind)
+            {
+                Debug.Assert(
+                    nodeToBind.Kind() == SyntaxKind.CasePatternSwitchLabel ||
+                    nodeToBind.Kind() == SyntaxKind.ArgumentList && nodeToBind.Parent is ConstructorInitializerSyntax ||
+                    nodeToBind is ExpressionSyntax);
```

Comment:
Consider moving or duplicating this assert in the factory method, replacing ```Debug.Assert(nodeToBind.Kind() != SyntaxKind.BracketedArgumentList);``` with it. The benefit would be to checking for things we expect rather than for things we don't. In addition, the assumption would be validated for more scenarios, this code path is only taken for 'var' case.